### ISSUE TITLE
Inconsistent tf.gather_nd behavior on cpu vs gpu

### ISSUE THREAD

**Issue by cheyennee** [tensorflow/tensorflow#57170]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

TF2.4

### Custom Code

Yes

### Current Behaviour?

```shell
The result on cpu should be consistent with the result on gpu.
```

### Standalone code to reproduce the issue

```shell
results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_tensor = tf.random.uniform([1, 8, 1, 1], dtype=tf.float32)
      arg_0 = tf.identity(arg_0_tensor)
      arg_1_tensor = tf.saturate_cast(tf.random.uniform([8, 2], minval=-256, maxval=257, dtype=tf.int64), dtype=tf.int64)
      arg_1 = tf.identity(arg_1_tensor)
      results["res_cpu"] = tf.gather_nd(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = tf.identity(arg_0_tensor)
      arg_0 = tf.cast(arg_0, tf.float32)
      arg_1 = tf.identity(arg_1_tensor)
      arg_1 = tf.cast(arg_1, tf.int64)
      results["res_gpu"] = tf.gather_nd(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)

print(results)
'''
{'err_cpu': 'Error:indices[7] = [-232, -191] does not index into param shape [1,8,1,1] [Op:GatherNd]', 'res_gpu': <tf.Tensor: shape=(8, 1, 1), dtype=float32, numpy=
array([[[0.]],
       [[0.]],
       [[0.]],
       [[0.]],
       [[0.]],
       [[0.]],
       [[0.]],
       [[0.]]], dtype=float32)>}
'''
```

### Relevant log output

_No response_</details>

**Comment by cheyennee** [comment]
Same problem is also found in tf.gather.
```
results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_tensor = tf.random.uniform([1, 8, 1, 1], dtype=tf.float32)
      arg_0 = tf.identity(arg_0_tensor)
      arg_1_tensor = tf.saturate_cast(tf.random.uniform([1, 222], minval=-256, maxval=257, dtype=tf.int64), dtype=tf.int32)
      arg_1 = tf.identity(arg_1_tensor)
      results["res_cpu"] = tf.gather(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = tf.identity(arg_0_tensor)
      arg_0 = tf.cast(arg_0, tf.float32)
      arg_1 = tf.identity(arg_1_tensor)
      arg_1 = tf.cast(arg_1, tf.int32)
      results["res_gpu"] = tf.gather(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)
print(results)
```
The results is {'err_cpu': 'Error:indices[0,0] = 76 is not in [0, 1) [Op:GatherV2]', 'res_gpu': <tf.Tensor: shape=(1, 222, 8, 1, 1), dtype=float32, numpy=
array([[[[[0.]],
         [[0.]],
         [[0.]],
         ...,
         [[0.]],
         [[0.]],
         [[0.]]],
        [[[0.]],
         [[0.]],
         [[0.]],
         ...,
         [[0.]],
         [[0.]],
         [[0.]]],
        [[[0.]],
         [[0.]],
         [[0.]],
         ...,
         [[0.]],
         [[0.]],
         [[0.]]],
        ...,
        [[[0.]],
         [[0.]],
         [[0.]],
         ...,
         [[0.]],
         [[0.]],
         [[0.]]],
        [[[0.]],
         [[0.]],
         [[0.]],
         ...,
         [[0.]],
         [[0.]],
         [[0.]]],
        [[[0.]],
         [[0.]],
         [[0.]],
         ...,
         [[0.]],
         [[0.]],
         [[0.]]]]], dtype=float32)>}

**Comment by cheyennee** [comment]
Problem is found in tf.keras.losses.sparse_categorical_crossentropy.
```
results = dict()
import tensorflow as tf
try:
  try:
    with tf.device('/CPU'):
      arg_0_0 = -56
      arg_0_1 = -47
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0_0 = 0.05
      arg_1_0_1 = 0.95
      arg_1_0_2 = 0
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1_0 = 0.1
      arg_1_1_1 = 0.8
      arg_1_1_2 = 0.1
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_cpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_cpu"] = "Error:"+str(e)
  try:
    with tf.device('/GPU:0'):
      arg_0 = [arg_0_0,arg_0_1,]
      arg_1_0 = [arg_1_0_0,arg_1_0_1,arg_1_0_2,]
      arg_1_1 = [arg_1_1_0,arg_1_1_1,arg_1_1_2,]
      arg_1 = [arg_1_0,arg_1_1,]
      results["res_gpu"] = tf.keras.losses.sparse_categorical_crossentropy(arg_0,arg_1,)
  except Exception as e:
    results["err_gpu"] = "Error:"+str(e)
except Exception as e:
  results["err"] = "Error:"+str(e)
print(results)
```
The results is {'err_cpu': 'Error:Received a label value of -56 which is outside the valid range of [0, 3).  Label values: -56 -47 [Op:SparseSoftmaxCrossEntropyWithLogits]', 'res_gpu': <tf.Tensor: shape=(2,), dtype=float32, numpy=array([nan, nan], dtype=float32)>}. 
CPU throws an error, while GPU runs.

**Comment by sushreebarsa** [comment]
@cheyennee I tried to replicate the issue on colab, please find the gist here in [cpu](https://colab.research.google.com/gist/sushreebarsa/f4e4093822f352682d402e7581fc8552/57170-cpu.ipynb) and [gpu](https://colab.research.google.com/gist/sushreebarsa/eb93c6fc6c40cec0d1c7f332678121ae/57170-gpu.ipynb) runtime. Could you please confirm the issue. Thank you!

**Comment by cheyennee** [comment]
@sushreebarsa Are the two codes based on different tensorflow versions? Later version may have fixed bugs from previous version. Otherwise it's hard to explain why the same code behaves differently😶‍🌫️

**Comment by sushreebarsa** [comment]
@cheyennee The gists are having latest TF versions. Could you please check if this issue is not replicating in latest versions as it is recommended to use latest TF version. Thank you!

**Comment by cheyennee** [comment]
@sushreebarsa I have confirmed that these issues are not replicating in tf 2.8.2 version. I will update my tf version. Thanks.
